src/patterns.md

```diff
@@ -0,0 +1,656 @@
+```
+
+The grammar production for reference patterns has to match the token `&&` to match a
+reference to a reference because it is a token by itself, not two `&` tokens.
+
+Adding the `mut` keyword dereferences a mutable reference. The mutability must match the
+mutability of the reference.
+
+Reference patterns are always irrefutable.
+
+## Struct patterns
+
+> **<sup>Syntax</sup>**\
+> _StructPattern_ :\
+> &nbsp;&nbsp; _Path_ `{`\
+> &nbsp;&nbsp; &nbsp;&nbsp; _StructPatternElements_ <sup>?</sup>\
+> &nbsp;&nbsp; `}`
+>
+> _StructPatternElements_ :\
+> &nbsp;&nbsp; &nbsp;&nbsp; _StructPatternFields_ (`,` | `,` _StructPatternEtCetera_)<sup>?</sup>\
+> &nbsp;&nbsp; | _StructPatternEtCetera_
+>
+> _StructPatternFields_ :\
+> &nbsp;&nbsp; _StructPatternField_ (`,` _StructPatternField_) <sup>\*</sup>
+>
+> _StructPatternField_ :\
+> &nbsp;&nbsp; [_OuterAttribute_] <sup>\*</sup>\
+> &nbsp;&nbsp; (\
+> &nbsp;&nbsp; &nbsp;&nbsp; &nbsp;&nbsp; [INTEGER_LITERAL] `:` [_Pattern_]\
+> &nbsp;&nbsp; &nbsp;&nbsp; | [IDENTIFIER] `:` [_Pattern_]\
+> &nbsp;&nbsp; &nbsp;&nbsp; | `ref`<sup>?</sup> `mut`<sup>?</sup> [IDENTIFIER]\
+> &nbsp;&nbsp; )
+>
+> _StructPatternEtCetera_ :\
+> &nbsp;&nbsp; [_OuterAttribute_] <sup>\*</sup>\
+> &nbsp;&nbsp; `..`
+
+[_OuterAttribute_]: attributes.html
+
+Struct patterns match struct values that match all criteria defined by its subpatterns.
+They are also used to [destructure](#destructuring) a struct.
+
+On a struct pattern, the fields are referenced by name, index (in the case of tuple
+structs) or ignored by use of `..`:
+
+```rust
+# struct Point {
+#     x: u32,
+#     y: u32,
+# }
+# let s = Point {x: 1, y: 1};
+#
+match s {
+    Point {x: 10, y: 20} => (),
+    Point {y: 10, x: 20} => (),    // order doesn't matter
+    Point {x: 10, ..} => (),
+    Point {..} => (),
+}
+
+# struct PointTuple (
+#     u32,
+#     u32,
+# );
+# let t = PointTuple(1, 2);
+#
+match t {
+    PointTuple {0: 10, 1: 20} => (),
+    PointTuple {1: 10, 0: 20} => (),   // order doesn't matter
+    PointTuple {0: 10, ..} => (),
+    PointTuple {..} => (),
+}
+```
+
+If `..` is not used, it is required to match all fields:
+
+```rust
+# struct Struct {
+#    a: i32,
+#    b: char,
+#    c: bool,
+# }
+# let mut struct_value = Struct{a: 10, b: 'X', c: false};
+#
+match struct_value {
+    Struct{a: 10, b: 'X', c: false} => (),
+    Struct{a: 10, b: 'X', ref c} => (),
+    Struct{a: 10, b: 'X', ref mut c} => (),
+    Struct{a: 10, b: 'X', c: _} => (),
+    Struct{a: _, b: _, c: _} => (),
+}
+```
+
+The `ref` and/or `mut` _IDENTIFIER_ syntax matches any value and binds it to
+a variable with the same name as the given field.
+
+```rust
+# struct Struct {
+#    a: i32,
+#    b: char,
+#    c: bool,
+# }
+# let struct_value = Struct{a: 10, b: 'X', c: false};
+#
+let Struct{a: x, b: y, c: z} = struct_value;          // destructure all fields
+```
+
+A struct pattern is refutable when one of its subpatterns is refutable.
+
+## TupleStruct patterns
```

Comment:
Do we want "TupleStruct" to be one word or two?

I generally don't like smashing words together.